### PR TITLE
🚀 Upgrade to Lightning 2.0

### PR DESCRIPTION
**Pull request by alafage**
Start from TorchUncertainty 0.1.6:
- [x] Merge modifications from main

Transition to Lightning CLI:
- [x] rework classification experiment folder
- [x] rework regression experiment folder
- [x] create a segmentation experiment folder

Routine simplification:
- [x] ClassificationRoutine (add `save_in_csv` parameter to write the results in a csv file `on_test_epoch_end()`)
- [x] RegressionRoutine
- [x] SegmentationRoutine

Baseline update:
- [x] ResNet
- [x] VGG
- [x] WideResNet
- [x] Deep Ensembles (classification)
- [x] MLP
- [x] Deep Ensembles (regression)
- [x] Handle ELBO Loss
- [x] Handle Normal Inverse Gamma (create a new torch distribution)

Datamodules update:
- [x] MNIST
- [x] CIFAR10
- [x] CIFAR100
- [x] TinyImageNet
- [x] ImageNet

New datamodules:
- [x] Cityscapes
- [x] MUAD

Major issues:
- [ ] Cross-validation using LightningCLI: a simple solution would be to set `LightningCLI(run=False)` and call the fit method of the trainer ourselves as in the previous `cli_main()`. -- Moved to later

Final steps:
- [x] Delete useless files
- [ ] Fix coverage
- [x] Complete classification tutorials overhaul
- [x] Evidential regression tutorial
- [ ] Fix all docstrings
- [x] Switch to 0.2.0

Dependencies:
- [x] Remove cli-test-helpers (#81 #82)

### CODE REVIEW

File: experiments/classification/cifar10/readme.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+python resnet.py fit --config configs/resnet50/packed.yaml
+```
+
+
+**Note:** In addition we provide a default resnet config file (`configs/resnet.yaml`) to enable the training of any ResNet model. Here a basic example to train a MIMO ResNet101 model with $4$ estimators and $\rho=1.0$:
+
+```bash
+python resnet.py fit --config configs/resnet.yaml --model.arch 101 --model.version mimo --model.num_estimators 4 --model.rho 1.0
+```
+
```
Comment (o-laurent) [inline experiments/classification/cifar10/readme.md:29]: Too many line breaks?

Comment (o-laurent) [inline experiments/classification/cifar10/readme.md:29]: Too many line breaks?

File: experiments/classification/cifar100/readme.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# CIFAR100 - Benchmark
+
+TODO
```
Comment (o-laurent) [inline experiments/classification/cifar100/readme.md:3]: Useless for now

File: experiments/classification/mnist/bayesian_lenet.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def optim_lenet(model: nn.Module) -> dict:
         num_classes=dm.num_classes,
         in_channels=dm.num_channels,
         loss=loss,
-        optimization_procedure=optim_lenet,
+        optim_recipe=optim_lenet,
```
Comment (o-laurent) [inline experiments/classification/mnist/bayesian_lenet.py:58]: Not updated + this doesn't work anymore

File: experiments/classification/mnist/lenet.py
Comment (o-laurent) [inline experiments/classification/mnist/lenet.py:1]: not updated

File: experiments/classification/tiny-imagenet/resnet.py
Comment (o-laurent) [inline experiments/classification/tiny-imagenet/resnet.py:1]: not updated & broken

File: experiments/regression/uci_datasets/deep_ensemble.py
Comment (o-laurent) [inline experiments/regression/uci_datasets/deep_ensemble.py:1]: not updated & broken

File: experiments/segmentation/readme.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# Segmentation Benchmarks
```
Comment (o-laurent) [inline experiments/segmentation/readme.md:1]: not filled

File: tests/baselines/test_mimo.py [hunk 5]
```diff
@@ -2,23 +2,26 @@
...
-    optim_cifar10_resnet50,
-    optim_cifar10_wideresnet,
+from torch_uncertainty.baselines.classification import (
+    ResNetBaseline,
+    WideResNetBaseline,
 )
 
+# from torch_uncertainty.optim_recipes import (
+#     optim_cifar10_resnet18,
+#     optim_cifar10_resnet50,
```
Comment (o-laurent) [inline tests/baselines/test_mimo.py:12]: delete

File: torch_uncertainty/routines/classification.py [hunk 6]
```diff
@@ -81,72 +85,53 @@ def __init__(
...
-            use_logits (bool, optional): Indicates whether to use the logits as the
-                OOD criterion or not. Defaults to ``False``.
+            ood_criterion (str, optional): OOD criterion. Defaults to ``"msp"``.
+                MSP is the maximum softmax probability, logit is the maximum
+                logit, energy the logsumexp of the mean logits, entropy the
+                entropy of the mean prediction, mi is the mutual information of
+                the ensemble and vr is the variation ratio of the ensemble.
             log_plots (bool, optional): Indicates whether to log plots from
                 metrics. Defaults to ``False``.
+            save_in_csv(bool, optional): __TODO__
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:95]: add docstring

File: torch_uncertainty/baselines/segmentation/segformer.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                Determines which architecture to use. Options are:
+
+                - ``0``: SegFormer-B0
+                - ``1``: SegFormer-B1
+                - ``2``: SegFormer-B2
+                - ``3``: SegFormer-B3
+                - ``4``: SegFormer-B4
+                - ``5``: SegFormer-B5
+
+            num_estimators (int, optional): _description_. Defaults to 1.
```
Comment (o-laurent) [inline torch_uncertainty/baselines/segmentation/segformer.py:59]: description missing

File: torch_uncertainty/losses.py [hunk 8]
```diff
@@ -34,7 +66,7 @@ def _kl_div(self) -> Tensor:
 class ELBOLoss(nn.Module):
     def __init__(
         self,
-        model: nn.Module,
+        model: nn.Module | None,
```
Comment (o-laurent) [inline torch_uncertainty/losses.py:69]: Add docstring explanation

File: torch_uncertainty/metrics/nll.py [hunk 9]
```diff
@@ -2,14 +2,15 @@
...
 from torchmetrics import Metric
 from torchmetrics.utilities.data import dim_zero_cat
 
 
-class NegativeLogLikelihood(Metric):
-    is_differentiable: bool = False
-    higher_is_better: bool | None = False
-    full_state_update: bool = False
+class CategoricalNLL(Metric):
+    is_differentiabled = False
```
Comment (o-laurent) [inline torch_uncertainty/metrics/nll.py:11]: change to is_differentiable

**Review by o-laurent — CHANGES_REQUESTED** [review]
